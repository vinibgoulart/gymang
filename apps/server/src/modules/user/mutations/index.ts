import UserAdd from './UserAddMutation';
import UserLogin from './UserLoginMutation';
import UserLogout from './UserLogoutMutation';

const UserMutations = {
  UserAdd,
  UserLogin,
  UserLogout,
};

export default UserMutations;
