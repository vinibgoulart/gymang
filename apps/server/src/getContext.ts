import { GRAPHQL_TYPE } from '@gymang/core';
import { getDataloaders } from './loader/loaderRegistry';
import { UserLoader } from '@gymang/user';

export const getContext = async (ctx: Record<string, unknown>) => {
  let context = {
    ...ctx,
  };

  const dataloaders = getDataloaders();

  const graphql = ctx.graphql || GRAPHQL_TYPE.WEB;

  if (context.user) {
    context.user = await UserLoader.load(
      { ...context, dataloaders, graphql: GRAPHQL_TYPE.WEB },
      context.user._id,
      true,
    );
  }

  const koaContext = {
    cookies: {
      // eslint-disable-next-line
      set: () => {},
    },
  };

  return {
    dataloaders,
    graphql,
    koaContext,
    t: (key: string) => key,
    ...context,
  };
};
