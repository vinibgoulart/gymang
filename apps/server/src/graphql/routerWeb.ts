import Router from "@koa/router";
import {
  getGraphQLParameters,
  processRequest,
  renderGraphiQL,
  shouldRenderGraphiQL,
} from "graphql-helix";
import { koaPlayground } from "graphql-playground-middleware";
import type koa from "koa";

import { getAuth } from "./auth/getAuth";
import { schema } from "./schema/schema";
import { getDataloaders } from "../loader/loaderRegistry";
import { config } from "@gymang/config";
import { GRAPHQL_TYPE } from "@gymang/core";

export const serverHelix = async (ctx: koa.Context, next: koa.Next) => {
  const request = ctx.request;

  if (shouldRenderGraphiQL(request) && config.GYMANG_ENV !== "production") {
    ctx.body = renderGraphiQL({ endpoint: "/admin" });
    return;
  }

  // eslint-disable-next-line
  console.log("operationName", request.body.operationName);

  const dataloaders = ctx?.dataloaders ?? getDataloaders();

  const context = {
    dataloaders,
    graphql: GRAPHQL_TYPE.WEB,
    koaContext: ctx,
    user: ctx.user,
    // @todo implement i18n with react-i18next
    t: (key: string) => key,
  };

  const test = getGraphQLParameters(request);
  const { operationName, query, variables } = test;

  const result = await processRequest({
    operationName,
    query,
    variables,
    request,
    schema,
    contextFactory: () => context,
  });

  if (result.type === "RESPONSE") {
    result.headers.forEach(({ name, value }) => {
      ctx.headers[name] = value;
    });
    ctx.status = result.status;
    ctx.body = result.payload;
    await next();
    return;
  }

  ctx.status = 200;
  ctx.headers = {
    "Content-Type": "text/event-stream",
    Connection: "keep-alive",
    "Cache-Control": "no-cache",
  };
  await next();
};

export const routerWeb = new Router();

routerWeb.all("/admin", getAuth);

routerWeb.all(
  "/admin/playground",
  koaPlayground({
    endpoint: "/admin",
  })
);

routerWeb.all("/admin", serverHelix);
