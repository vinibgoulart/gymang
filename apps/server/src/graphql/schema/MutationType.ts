import { GraphQLObjectType } from 'graphql';

import ExerciseMutations from '../../modules/exercise/mutations';
import UserMutations from '../../modules/user/mutations';
import WorkoutMutations from '../../modules/workout/mutations';
import WorkoutSplitMutations from '../../modules/workoutSplit/mutations';

const MutationType = new GraphQLObjectType({
  name: 'Mutation',
  fields: () => ({
    ...UserMutations,
    ...WorkoutMutations,
    ...WorkoutSplitMutations,
    ...ExerciseMutations,
  }),
});

export default MutationType;
