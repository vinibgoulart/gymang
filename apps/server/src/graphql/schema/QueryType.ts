import type { GraphQLContext } from '@gymang/core';
import { nodeField, nodesField } from '@gymang/graphql';
import { GraphQLObjectType } from 'graphql';
import { globalIdField } from 'graphql-relay';

import UserQueries from '../../modules/user/UserQueries';
import WorkoutQueries from '../../modules/workout/WorkoutQueries';
import WorkoutSplitQueries from '../../modules/workoutSplit/WorkoutSplitQueries';

const QueryType = new GraphQLObjectType<
  Record<string, unknown>,
  GraphQLContext
>({
  name: 'Query',
  description: 'The root of all... queries',
  fields: () => ({
    id: globalIdField('Query'),
    node: nodeField,
    nodes: nodesField,
    ...UserQueries,
    ...WorkoutQueries,
    ...WorkoutSplitQueries,
  }),
});

export default QueryType;
