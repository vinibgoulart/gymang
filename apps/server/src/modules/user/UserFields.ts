import { GraphQLNonNull } from 'graphql';
import UserType, { UserConnection } from './UserType';
import { UserLoader } from '@gymang/user';
import {
  edgeField,
  connectionArgs,
  NullConnection,
  withFilter,
} from '@gymang/graphql';
import { isLoggedIn } from '@gymang/core';

export const userTypeField = (key = 'user', bypassViewerCanSee = false) => ({
  [key]: {
    type: UserType,
    resolve: async (obj, args, context) => {
      return UserLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const userEdgeField = () =>
  edgeField({
    connection: UserConnection,
    load: UserLoader.load,
    edgeName: 'userEdge',
    name: 'User',
  });

export const userConnectionField = (customResolver = null) => ({
  users: {
    type: new GraphQLNonNull(UserConnection.connectionType),
    args: {
      ...connectionArgs,
      // filters: {
      //   type: UserFilterInputType,
      // },
    },
    resolve: (obj, args, context) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      const argsWithCompany = withFilter(args, {
        merchant: context?.merchant?._id,
      });

      return UserLoader.loadAll(context, argsWithCompany);
    },
  },
});

export const meField = () => ({
  me: {
    type: UserType,
    resolve: (_root, _args, context) => {
      return UserLoader.load(context, context.user?._id);
    },
  },
});
