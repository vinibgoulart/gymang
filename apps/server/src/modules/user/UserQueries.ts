import { meField, userConnectionField } from './UserFields';

const UserQueries = {
  ...meField(),
  ...userConnectionField(),
};

export default UserQueries;
