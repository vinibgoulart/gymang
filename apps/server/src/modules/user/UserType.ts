import { GraphQLContext } from '@gymang/core';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import { IUser, UserLoader } from '@gymang/user';
import { GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

const UserType = new GraphQLObjectType<IUser, GraphQLContext>({
  name: 'User',
  description: 'Represents a User',
  fields: () => ({
    id: globalIdField('User'),
    firstName: {
      type: GraphQLString,
      resolve: (user) => user.firstName,
    },
    email: {
      type: GraphQLString,
      resolve: (user) => user.email,
    },
  }),
  interfaces: () => [nodeInterface],
});

export const UserConnection = connectionDefinitions({
  name: 'User',
  nodeType: UserType,
});

registerTypeLoader(UserType, UserLoader.load);

export default UserType;
