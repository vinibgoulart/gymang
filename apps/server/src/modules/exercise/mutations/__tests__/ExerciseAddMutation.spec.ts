import { GRAPHQL_TYPE } from '@gymang/core';
import { MUSCLE_GROUP } from '@gymang/enums';
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from '@gymang/testutils';
import { handleCreateUser } from '@gymang/user';
import { handleCreateWorkoutSplit } from '@gymang/workout-split';
import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import { getContext } from '../../../../getContext';
import { schema as schemaWeb } from '../../../../graphql/schema/schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const query = `
  mutation ExerciseAddSpecMutation($input: ExerciseAddInput!) {
    ExerciseAdd(input: $input) {
      exercise {
        id
        name
        user {
          firstName
        }
        workoutSplit {
          id
          name
        }
        series
        repetitions
        breakTime
        muscleGroup
        weight
      }
      success
      error
    }
  }
`;

it('should add a new exercise', async () => {
  const user = await handleCreateUser();
  const workoutSplit = await handleCreateWorkoutSplit();

  const input = {
    name: 'Pull Down',
    workoutSplit: toGlobalId('WorkoutSplit', workoutSplit._id),
    series: '3',
    repetitions: '10',
    weight: '50',
    breakTime: '60',
    muscleGroup: MUSCLE_GROUP.BACK,
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.ExerciseAdd.error).toBeNull();
  expect(result.data.ExerciseAdd.success).toEqual(
    'Exercise created successfully',
  );

  const exerciseCreated = result.data.ExerciseAdd.exercise;

  expect(exerciseCreated?.user.firstName).toEqual(user.firstName);
  expect(exerciseCreated?.workoutSplit.name).toEqual(workoutSplit.name);

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it('should not add a new exercise for another user', async () => {
  const user = await handleCreateUser();
  const anotherUser = await handleCreateUser();
  const workoutSplit = await handleCreateWorkoutSplit({
    user: anotherUser,
  });

  const input = {
    name: 'Pull Down',
    workoutSplit: toGlobalId('WorkoutSplit', workoutSplit._id),
    series: '3',
    repetitions: '10',
    weight: '50',
    breakTime: '60',
    muscleGroup: MUSCLE_GROUP.BACK,
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.ExerciseAdd.error).toEqual(
    'You can not create a Exercise for another user',
  );
  expect(result.data.ExerciseAdd.success).toBeNull();
});
