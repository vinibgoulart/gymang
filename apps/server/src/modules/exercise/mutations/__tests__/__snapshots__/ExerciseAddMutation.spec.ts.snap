// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a new exercise 1`] = `
{
  "ExerciseAdd": {
    "error": "EMPTY",
    "exercise": {
      "breakTime": "60",
      "id": "FROZEN-ID",
      "muscleGroup": "Back",
      "name": "Pull Down",
      "repetitions": "10",
      "series": "3",
      "user": {
        "firstName": "User 1 firstName",
      },
      "weight": "50",
      "workoutSplit": {
        "id": "FROZEN-ID",
        "name": "WorkoutSplit NaN name",
      },
    },
    "success": "Exercise created successfully",
  },
}
`;
